pkg/proxy/userspace/proxier.go

```diff
@@ -557,22 +595,83 @@ func (proxier *Proxier) unmergeService(service *v1.Service, existingPorts sets.S
 	}
 }
 
+// If a change exists for this service name, copies the service information
+// to the change and returns it. If a change does not yet exist, creates a
+// new change with the given event type, copies the service information to it,
+// and returns it.
+func (proxier *Proxier) ensureServiceChange(service *v1.Service, event serviceEventType) (*serviceChange, types.NamespacedName) {
+	svcName := types.NamespacedName{Namespace: service.Namespace, Name: service.Name}
+	change, exists := proxier.serviceChanges[svcName]
+	if !exists {
+		change = &serviceChange{event: event}
+		proxier.serviceChanges[svcName] = change
+	}
+	change.service = service.DeepCopy()
+	return change, svcName
+}
+
 func (proxier *Proxier) OnServiceAdd(service *v1.Service) {
-	_ = proxier.mergeService(service)
+	klog.V(4).Infof("userspace proxy: OnServiceAdd for %s/%s", service.Namespace, service.Name)
+
+	proxier.serviceChangesLock.Lock()
+	defer proxier.serviceChangesLock.Unlock()
+
+	change, _ := proxier.ensureServiceChange(service, serviceEventAdd)
+	if change.event == serviceEventDelete {
+		// Delete + Add = Update
+		change.event = serviceEventUpdate
+	}
+
+	proxier.syncRunner.Run()
 }
 
 func (proxier *Proxier) OnServiceUpdate(oldService, service *v1.Service) {
-	existingPorts := proxier.mergeService(service)
-	proxier.unmergeService(oldService, existingPorts)
+	if service == nil {
+		// Only care about the most up-to-date service, because the proxy
+		// caches any previous version of this service for cleanup
+		return
+	}
+	klog.V(4).Infof("userspace proxy: OnServiceUpdate for %s/%s", service.Namespace, service.Name)
+
+	proxier.serviceChangesLock.Lock()
+	defer proxier.serviceChangesLock.Unlock()
+
+	change, _ := proxier.ensureServiceChange(service, serviceEventUpdate)
+	if change.event == serviceEventDelete {
+		// Delete + Update = Update
+		change.event = serviceEventUpdate
+	}
+
+	proxier.syncRunner.Run()
 }
 
 func (proxier *Proxier) OnServiceDelete(service *v1.Service) {
-	proxier.unmergeService(service, sets.NewString())
+	klog.V(4).Infof("userspace proxy: OnServiceDelete for %s/%s", service.Namespace, service.Name)
+
+	proxier.serviceChangesLock.Lock()
+	defer proxier.serviceChangesLock.Unlock()
+
+	change, svcName := proxier.ensureServiceChange(service, serviceEventDelete)
+	if change.event == serviceEventAdd {
+		// Add + Delete = ignored
+		delete(proxier.serviceChanges, svcName)
+		return
+	} else if change.event == serviceEventUpdate {
+		// Update + Delete = Delete
+		change.event = serviceEventDelete
+	}
+
+	proxier.syncRunner.Run()
 }
 
 func (proxier *Proxier) OnServiceSynced() {
 	klog.V(2).Infof("userspace OnServiceSynced")
 	proxier.setInitialized(true, false)
+
+	// Must sync from a goroutine to avoid blocking the
+	// service event handler on startup with large numbers
+	// of initial objects
+	go proxier.syncProxyRules()
```

Comment:
why not just `proxier.syncRunner.Run()`? (which also ensures that it runs from a goroutine, not the calling thread)
likewise in `OnEndpointsAdd`

Reply:
@danwinship because proxier.syncRunner.Run() respects the ratelimiting, which we don't want here. That said until we are initialized we shouldn't ever end up calling syncRunner.Run() so maybe it's OK to call Run() from here right after initializing, I guess.

Reply:
Turns out calling the syncrunner here does have an effect, it doubles the time the testcase takes to run. I'd rather be explicit about what we want here and not rely on the sync runner to "do the right thing" in this case because we know it might not given its purpose in life.